include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,103 +202,138 @@ struct __init_processing
+        for (_Size __i = 0; __i < _VecSize; ++__i)
+            __res.__v = __binary_op(__res.__v, __unary_op(__start_idx + __i, __acc...));
+    }
+
+    template <typename _Size, typename _Res, typename... _Acc>
+    void
+    scalar_reduction_remainder(const _Size __start_idx, const _Size __adjusted_n, const _Size __max_iters, _Res& __res,
+                               const _Acc&... __acc) const
+    {
+        const _Size __remainder = __adjusted_n - __start_idx;
+        const _Size __no_iters = std::min(__remainder, __max_iters);
+        for (_Size __idx = 0; __idx < __no_iters; ++__idx)
+            __res.__v = __binary_op(__res.__v, __unary_op(__start_idx + __idx, __acc...));
+    }
 
-        const _Size __adjusted_global_id =
-            __global_offset + __item_id.get_group_linear_id() * __stride * __iters_per_work_item + __local_idx;
+    template <typename _NDItemId, typename _Size, typename _Res, typename... _Acc>
+    void
+    operator()(const _NDItemId& __item_id, const _Size& __n, const _Size& __iters_per_work_item,
+               const _Size& __global_offset, const bool __is_full, const _Size __n_groups, _Res& __res,
+               const _Acc&... __acc) const
+    {
+        const _Size __global_idx = __item_id.get_global_id(0);
+        if (__iters_per_work_item == 1)
+        {
+            new (&__res.__v) _Tp(__unary_op(__global_idx, __acc...));
+            return;
+        }
+        const _Size __local_range = __item_id.get_local_range(0);
+        const _Size __no_vec_ops = __iters_per_work_item / _VecSize;
         const _Size __adjusted_n = __global_offset + __n;
+        constexpr _Size __vec_size_minus_one = _VecSize - 1;
 
-        // Coalesced load and reduce from global memory
-        if (__adjusted_global_id + __stride * __iters_per_work_item < __adjusted_n)
+        _Size __stride = _VecSize; // sequential loads with _VecSize-wide vectors
+        _Size __adjusted_global_id = __global_offset;
+        if constexpr (_Commutative{})
         {
-            new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            _ONEDPL_PRAGMA_UNROLL
-            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            __stride *= __local_range; // coalesced loads with _VecSize-wide vectors
+            _Size __local_idx = __item_id.get_local_id(0);
+            _Size __group_idx = __item_id.get_group_linear_id();
+            __adjusted_global_id += __group_idx * __local_range * __iters_per_work_item + __local_idx * _VecSize;
+        }
+        else
+            __adjusted_global_id += __iters_per_work_item * __global_idx;
+
+        // If n is not evenly divisible by the number of elements processed per work-group, the last work-group might
+        // need to process less elements than __iters_per_work_item.
+        bool __is_full_wg = __is_full;
```

Comment:
Can we simplify this to something like:
```
bool __is_full_wg = __is_full ||
    (static_cast<_Size>( __item_id.get_group(0)) < static_cast<_Size>(__n_groups - 1));
```
I think the `__is_multi_group` check is not needed since if there is one group and it is is full, then `__is_full` will be true already.

Reply:
I've changed the check and tried to clarify the comment.